docs/faas/aws-lambda.md

```diff
@@ -54,21 +54,14 @@ and the [cloud resource conventions][cloud]. The following AWS Lambda-specific a
 [faasres]: /docs/resource/faas.md (FaaS resource conventions)
 [cloud]: /docs/resource/cloud.md (Cloud resource conventions)
 
-### AWS X-Ray Environment Span Link
+### AWS X-Ray Environment `SpanContext`
 
-If the `_X_AMZN_TRACE_ID` environment variable is set, instrumentation SHOULD try to parse an
-OpenTelemetry `Context` out of it using the [AWS X-Ray Propagator](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.22.0/specification/context/api-propagators.md). If the
-resulting `Context` is [valid](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.22.0/specification/trace/api.md#isvalid) then a [Span Link][] SHOULD be added to the new Span's
-[start options](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.22.0/specification/trace/api.md#specifying-links) with an associated attribute of `source=x-ray-env` to
-indicate the source of the linked span.
-Instrumentation MUST check if the context is valid before using it because the `_X_AMZN_TRACE_ID` environment variable can
-contain an incomplete trace context which indicates X-Ray isn’t enabled. The environment variable will be set and the
-`Context` will be valid and sampled only if AWS X-Ray has been enabled for the Lambda function. A user can
-disable AWS X-Ray for the function if the X-Ray Span Link is not desired.
+If the `_X_AMZN_TRACE_ID` environment variable is set, instrumentation MUST ensure that the value from that
+environment variable is included in the `Carrier` used to extract a parent `Context` with the `Field` name
```

Comment:
I want to point out that as currently written, this document doesn't specify what should happen to `X-Amzn-Trace-Id` if it already exists in the carrier before the environment variable value is applied to it.  I think that's an important behavior to define.

Reply:
> Setting that value will override any value that may have existed.

what would be the downside for this to be added as a separate key in the carrier, e.g. `x-amzn-env-trace-id` or `x-amzn-vended-trace-id` (instead of overriding the existing carrier value), this way the propagator chain can have access to all of the different propagated trace contexts?

Reply:
> > Setting that value will override any value that may have existed.
> 
> what would be the downside for this to be added as a separate key in the carrier, e.g. `x-amzn-env-trace-id` or `x-amzn-vended-trace-id` (instead of overriding the existing carrier value), this way the propagator chain can have access to all of the different propagated trace contexts?

I believe that I explained why that is not appropriate [here](https://github.com/open-telemetry/semantic-conventions/pull/277#discussion_r1304622684).  

I have repeatedly said that I believe that there needs to be more user agency afforded with regard to the construction of the carrier used for context extraction.  Doing so would make it irrelevant what any of us think about what users want because they would be able to do what they want without configuration options or overrides.  I have avoided going that far in this PR as here I am trying to fix a broken spec in the smallest possible way so that instrumentation can start offering more flexibility to users in a way that leaves the door open for further changes in that direction.  Can we all agree that this change does that?

Reply:
> in the smallest possible way

I think the smallest possible way wouldn't change the current default behavior, and instead would introduce an opt-in behavior to overwrite the span context in the carrier with the span context from the env var.

Reply:
> It also doesn't give any option to users who wish to disable this behavior.

Why the option of using w3c is not enough? this proposal is not blocking users of using the context propagator of their choice. 

> what would be the downside for this to be added as a separate key in the carrier, e.g. x-amzn-env-trace-id or x-amzn-vended-trace-id 

The [lambda documentation states](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html#configuration-envvars-runtime) that `_X_AMZN_TRACE_ID` is where the [X-Ray tracing header](https://docs.aws.amazon.com/lambda/latest/dg/services-xray.html) is. So one downside would be the lack of consistency.

My two cents is that this is how context propagation using x-ray was designed to work in aws lambda and this is the current state. The lambda runtime plays an important role on it and there is currently no way to disable it. Do we really need to bend the x-ray trace headers to a different purpose from what it was designed when there are standard alternatives available to circunvent these limitations?

Reply:
AWS doesn't support propagating w3c in the same way it propagates x-ray.

Reply:
> AWS doesn't support propagating w3c in the same way it propagates x-ray.

@tylerbenson can you explain this?

Reply:
W3C can only propagate over http headers or sqs/sns message attributes. That's fine in simple architectures as long as users haven't added many of their own attributes. AWS imposes a limit of 10 message attributes which means w3c propagation can fail if the limit is already reached. X-Ray is not subject to this limit as it propagates over message _system_ attributes.
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-metadata.html#sqs-message-system-attributes

Attributes also don't allow propagation when using S3 events or SNS->SQS type architectures (among other scenarios involving aws managed services). Tracing in these scenarios require x-ray. This was also the main motivation for https://github.com/open-telemetry/opentelemetry-specification/pull/3212.

Reply:
> Attributes also don't allow propagation when using S3 events or SNS->SQS type architectures (among other scenarios involving aws managed services). Tracing in these scenarios require x-ray.

is there any AWS (or other) doc about this, I feel like this is an important point that I'd like to understand better

Reply:
Here are a few links I found:
https://docs.aws.amazon.com/sns/latest/dg/sns-active-tracing.html
https://aws.amazon.com/about-aws/whats-new/2020/11/aws-x-ray-supports-trace-context-propagation-amazon-simple-storage-service-s3/
https://docs.aws.amazon.com/xray/latest/devguide/xray-services-sqs.html

The key point being this all applies only to x-ray.

Reply:
Hi @jsuereth,

Now that you have context with this proposal, there is also this other proposal that tries to solve the problem with a different direction and that I believe has "yes" as answer for questions 1 and 2:  https://github.com/open-telemetry/semantic-conventions/pull/164

While I agree that @trask 's proposal would work, I would prefer to avoid it given that this propagator would be specific to lambda (i.e.: users would not be able to use this propagator elsewhere) and it would not fully comply with the propagator API, specifically:

> If a value can not be parsed from the carrier, for a cross-cutting concern, the implementation MUST NOT throw an exception and MUST NOT store a new value in the Context, in order to preserve any previously existing valid value.

Moreover, what would be the behaviour of the `inject` operation of this propagator?

I would prefer that the lambda instrumentation prepare a valid carrier to be consumed by subsequent propagators.